web-components/src/components/table-advanced/TableAdvanced.ts

```diff
@@ -0,0 +1,994 @@
+
+    private isResizing = false;
+    private isSelectable = false;
+    private selected: Record<number, boolean> = {};
+    private expandedRowIdx: Record<number, boolean> = {};
+
+    private dragover = (e: Event) => e.preventDefault();
+
+    disconnectedCallback() {
+      super.disconnectedCallback();
+      document.removeEventListener("dragover", this.dragover);
+    }
+
+    connectedCallback() {
+      super.connectedCallback();
+      document.addEventListener("dragover", this.dragover);
+
+      // cols
+
+      let index = 0;
+      const pushCol = (col: ColOptions, group?: { name: string; length: number }) => {
+        const f = col.filters || this.config.default?.col?.filters;
+        const filters = f
+          ? f == "forString"
+            ? Filter.optionsString
+            : f == "forNumber"
+              ? Filter.optionsNumber
+              : f.length
+                ? f
+                : null
+          : null;
+
+        this.COLS.push({
+          options: { ...col },
+          index: index++,
+          group,
+          sort: "default",
+          sorter: col.sorter || this.config.default?.col?.sorter,
+          isCollapsable: this.config.cols.collapse == col.id,
+          filter: filters
+            ? {
+              list: filters,
+              selectedIndex: 0,
+              input: "",
+              active: false,
+              menuVisible: false
+            }
+            : undefined
+        });
+      };
+
+      this.config.cols.define.forEach(col => {
+        if ("children" in col) {
+          col.children.forEach(c => pushCol(c, { name: col.groupName, length: col.children.length }));
+        } else {
+          pushCol(col);
+        }
+      });
+
+      const lenNodes = this.COLS.length;
+
+      // data
+
+      if ("csv" in this.data) {
+        const parse = Papa.parse(this.data.csv, {
+          skipEmptyLines: true,
+          transform: x => x.trim()
+        });
+        if (parse.errors) {
+          this.error = "PARSE ERROR:\n" + JSON.stringify(parse.errors, null, 2);
+          return;
+        } else {
+          const data = parse.data as string[][];
+          this.ROWS = data.map(x => x.map(text => ({ text })));
+        }
+      } else if ("list2d" in this.data) {
+        this.ROWS = this.data.list2d.map(x => x.map(text => ({ text })));
+      } else {
+        while (this.data.list.length > lenNodes) {
+          this.ROWS.push(this.data.list.splice(0, lenNodes).map(text => ({ text })));
+        }
+        if (this.data.list.length != 0) {
+          this.ROWS.push(this.data.list.map(text => ({ text })));
+        }
+      }
+
+      if (this.ROWS.length == 0) {
+        this.error = "DATA ERROR: Data is empty";
+        return;
+      }
+
+      // validate
+
+      const lenData = this.ROWS.reduce((acc, d) => acc + d.length, 0);
+      if (lenData % lenNodes != 0) {
+        this.error = "DATA ERROR: Dividing DATA by COLS is not zero";
```

Comment:
can you make this message a little more user friendly?

Reply:
done

Reply:
"DATA ERROR: Data length mismatch. You must provide (numberOfRows * numberOfColumns) amount of data values."

maybe someone can propose something less bone lingual?